core-client/transports/src/transports/ws.rs

```diff
@@ -77,65 +86,87 @@ where
 	}
 }
 
-impl<TSink, TStream, TError> Sink for WebsocketClient<TSink, TStream>
+impl<TSink, TStream> futures::Sink<String> for WebsocketClient<TSink, TStream>
 where
-	TSink: Sink<SinkItem = OwnedMessage, SinkError = TError>,
-	TStream: Stream<Item = OwnedMessage, Error = TError>,
-	TError: std::error::Error + Send + 'static,
+	TSink: futures::Sink<OwnedMessage, Error = RpcError> + Unpin,
+	TStream: futures::Stream<Item = Result<OwnedMessage, RpcError>> + Unpin,
 {
-	type SinkItem = String;
-	type SinkError = RpcError;
+	type Error = RpcError;
 
-	fn start_send(&mut self, request: Self::SinkItem) -> Result<AsyncSink<Self::SinkItem>, Self::SinkError> {
+	fn start_send(mut self: Pin<&mut Self>, request: String) -> Result<(), Self::Error> {
 		self.queue.push_back(OwnedMessage::Text(request));
-		Ok(AsyncSink::Ready)
+		Ok(())
 	}
 
-	fn poll_complete(&mut self) -> Result<Async<()>, Self::SinkError> {
+	fn poll_ready(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+		let this = Pin::into_inner(self);
+
 		loop {
-			match self.queue.pop_front() {
-				Some(request) => match self.sink.start_send(request) {
-					Ok(AsyncSink::Ready) => continue,
-					Ok(AsyncSink::NotReady(request)) => {
-						self.queue.push_front(request);
-						break;
+			match this.queue.pop_front() {
+				Some(request) => match Pin::new(&mut this.sink).poll_ready(cx) {
+					Poll::Ready(Ok(())) => {
+						if let Err(err) = Pin::new(&mut this.sink).start_send(request) {
```

Comment:
Right, so the way I wanted to structure it is that `poll_ready` is doing the heavy work and is responsible for draining our internal queue and gradually enqueueing the items in the underlying sink. As I understand, the previous behaviour was implemented to avoid greedily buffering until an explicit flush but rather to keep the queue as a fallback - if the sink is not yet ready, simply enqueue the item and process it later.

The more I think about it, the more I'm convinced we should just copy https://docs.rs/futures-util/0.3.8/src/futures_util/sink/buffer.rs.html :see_no_evil:

Reply:
I think I made a mistake and shouldn't really do (pseudocode):
```rust
fn poll_$variant(...) {
    if !queue.is_empty() {
        self.poll_ready(cx)
    else {
        underlying.poll_$variant(cx)
    }
}
```
but rather
```rust
fn poll_$variant(...) {
    if !queue.is_empty() {
        match self.poll_ready(cx) {
            Poll::Pending => return Poll::Pending,
            err @ Poll::Ready(Err(..)) => return err,
            ok => underlying.poll_$variant(cx)
    }
    else {
        underlying.poll_$variant(cx)
    }
}
```
because there may be a case where we might call `poll_flush` with a non-empty queue, then drain it in delegated `poll_ready` call but ultimately not guarantee that the underlying send will be completed (which is the point of `poll_flush`, as you rightfully noticed), because we only call `poll_ready` on the underlying sink, in the end.

I used the modified (unbounded) version of `futures_util::sink::Buffer` version in 94a4b232feeb708d7e10c64320efd7609081524c, which already contains the aforementioned fix.

Let me know if this looks good now!

Reply:
Looks good! I'm pretty sure that when #605 is implemented we will get rid of this code anyway so no need to do any more sophisticated refactorings.